Userland/Libraries/LibGUI/TextDocument.cpp

```diff
@@ -384,6 +384,99 @@ DeprecatedString TextDocument::text_in_range(TextRange const& a_range) const
+    TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(
+        Utf32View(line.code_points(), line.length()));
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    auto boundary_guess = boundaries_cache.at(0);
+
+    for (auto current_boundary_guess : boundaries_cache) {
+        // If the current boundary is further to the right, update our guess.
+        if (current_boundary_guess < cursor.column())
+            boundary_guess = current_boundary_guess;
+        else
+            break;
+    }
+
+    // Return the distance in code points from the cursor.
+    return boundary_guess;
+}
+
+// This function will return the position of the next grapheme cluster break,
+// relative to the cursor, for "correct looking" parsing of unicode based on
+// grapheme cluster boundary algorithm.
+size_t TextDocument::get_next_grapheme_cluster_boundary(
+    TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(
+        Utf32View(line.code_points(), line.length()));
+
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    auto boundary_guess = boundaries_cache.at(1);
+
+    for (auto current_boundary_guess : boundaries_cache.in_reverse()) {
+        // If the current boundary is further to the right, update our guess.
+        if (current_boundary_guess < cursor.column())
+            boundary_guess = current_boundary_guess;
+        else
+            break;
+    }
+    // Assume that the cursor cannot sit at a point that is not a boundary, so
+    // return previous guess.
+    return boundary_guess;
+}
+
+// This function will determine the number of codepoints of the glyph
+// immediately preceding the cursor, by determining if there is an emoji. If no
+// emoji is found it will return 1.
+size_t TextDocument::distance_between_cursor_and_previous_boundary(
+    TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    return cursor.column() - get_previous_grapheme_cluster_boundary(cursor);
+}
+
+// This function will determine the number of codepoints of the glyph
+// immediately following the cursor, by determining if there is an emoji. If no
+// emoji is found it will return 1.
+size_t TextDocument::distance_between_cursor_and_next_boundary(
+    TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    return cursor.column() + get_next_grapheme_cluster_boundary(cursor);
```

Comment:
This can't be right? If we're at column 10 and the next grapheme boundary is column 12, this returns 22. Please test this, and swap these around:
```c++
return get_next_grapheme_cluster_boundary(cursor) - cursor.column();
```

Reply:
Looking at existing code some more, I would also just get rid of these two functions (`distance_between_cursor_and_previous_boundary` and `distance_between_cursor_and_next_boundary`). The calling code in TextEditor.cpp does the math itself in all other cases, so let's make this match. i.e. the `else` branch of the `Key_Delete` case would become:
```
else {
    auto grapheme_break_position = document().get_next_grapheme_cluster_boundary(m_cursor);
    erase_count = grapheme_break_position - m_cursor.column();
}
```

Similar for the `Key_Backspace` case.

Reply:
Almost there, but neither of these comments are resolved :sweat_smile:

Reply:
Oh goodness I'm so sorry, I think I might be losing it. I just completely looked over this one..